cpp/dolfinx/mesh/Topology.cpp

```diff
@@ -342,4 +346,97 @@ void Topology::set_entity_permutation(std::size_t cell_n, int entity_dim,
 }
 //-----------------------------------------------------------------------------
 mesh::CellType Topology::cell_type() const { return _cell_type; }
+
+//-----------------------------------------------------------------------------
+std::tuple<Topology, std::vector<int>, std::vector<int>>
+mesh::create_topology(MPI_Comm comm,
+                      const graph::AdjacencyList<std::int64_t>& cells,
+                      const fem::ElementDofLayout& layout)
+{
+  const int size = dolfinx::MPI::size(comm);
+
+  // TODO: This step can be skipped for 'P1' elements
+
+  // Extract topology data, e.g.just the vertices. For P1 geometry this
+  // should just be the identity operator.For other elements the
+  // filtered lists may have 'gaps', i.e.the indices might not be
+  // contiguous.
+  const graph::AdjacencyList<std::int64_t> cells_v
+      = mesh::extract_topology(layout, cells);
+
+  // Compute the destination rank for cells on this process via graph
+  // partitioning
+  const std::vector<int> dest = PartitioningNew::partition_cells(
+      comm, size, layout.cell_type(), cells_v);
+
+  // Distribute cells to destination rank
+  const auto [my_cells, src, original_cell_index]
+      = PartitioningNew::distribute(comm, cells_v, dest);
+
+  // Build local cell-vertex connectivity, with local vertex indices
+  // [0, 1, 2, ..., n), from cell-vertex connectivity using global
+  // indices and get map from global vertex indices in 'cells' to the
+  // local vertex indices
+  auto [cells_local, local_to_global_vertices]
+      = PartitioningNew::create_local_adjacency_list(my_cells);
+
+  // Create (i) local topology object and (ii) IndexMap for cells, and
+  // set cell-vertex topology
+  Topology topology_local(layout.cell_type());
+  const int tdim = topology_local.dim();
+  auto map = std::make_shared<common::IndexMap>(comm, cells_local.num_nodes(),
+                                                std::vector<std::int64_t>(), 1);
+  topology_local.set_index_map(tdim, map);
+  auto _cells_local
+      = std::make_shared<graph::AdjacencyList<std::int32_t>>(cells_local);
+  topology_local.set_connectivity(_cells_local, tdim, 0);
+
+  const int n = local_to_global_vertices.size();
+  map = std::make_shared<common::IndexMap>(comm, n, std::vector<std::int64_t>(),
+                                           1);
+  topology_local.set_index_map(0, map);
+
+  // Create facets for local topology, and attach to the topology
+  // object. This will be used to find possibly shared cells
```

Comment:
Is this comment correct? Shared cells will come from the cell index map.

Reply:
It's the local topology, i.e. on a single rank, so the IndexMap is on `MPI_COMM_SELF` and is unaware of neighbours.

Reply:
OK, I just don't understand "possibly shared cells" here

Reply:
Create a 'local' Topology from the cells that I have. If a call is on the boundary of this local topology it might share a facet with another process.